core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java

```diff
@@ -412,13 +410,14 @@ private ChannelFilter(
             if (collectMinMax) {
                 minMaxComparison = blockTypeOperators.getComparisonUnorderedLastOperator(type);
             }
-            blockBuilder = type.createBlockBuilder(null, EXPECTED_BLOCK_BUILDER_SIZE);
+            int expectedSize = Math.min(maxDistinctValues, 8192) * 2;
```

Comment:
why `*2?`

Reply:
We don't stop precisely at `maxDistinctValues` as we check size of set at the end of ingesting a page, that's why gave it a bit more room to grow upfront.